Вторая версия мессенджера

Kosurij opened the pull request:
Добавлены:

Из функциональности:
 - Валидация форм на фокус / блур / сабмит;
 - Переход на SPA: все рендерится на стороне клиента;

Из инструментов:
 - TypeScript, линтеры;
 - Создан общий компонент и шина событий;
 - Внедрен компонентный подход.

Commit:
Создан базовый компонент Block и шина событий EventBus.
Страница логина переведена на компоненты.
Добавлен TS.

Commit:
Все страницы переведены на компоненты.

Commit:
Выполнена вся верстка

Commit:
Добавлена возможность работать с массивами для класса Block

Commit:
Исправлена типизация

Commit:
Добавлен роутинг

Commit:
Обновлены стили

Commit:
Обновлен роутинг для деплоя.

Commit:
Добавлена верстка переписки

Commit:
Внесен ряд правок.

Commit:
Добавлена валидация.

Commit:
Добавлена валидация текстового поля.

Commit:
Добавлен HttpRequest.ts

Commit:
Подключены линтеры

Commit:
Обновлен ридми

Commit:
Скорректирован ридми

gennady-bars commented on .eslintrc.json:
```diff
@@ -0,0 +1,103 @@
+{
```
* Страницы логина и регистрации тоже нужно обернуть в тег `main` по семантике. По семантике нужно помещать весь главный контент страницы в тег `main`.

gennady-bars commented on .eslintrc.json:
```diff
@@ -0,0 +1,103 @@
+{
+  "env": {
+    "browser": true,
+     "es6": true,
```
Лучше дать возможность поменять порт из консоли, если понадобится: `const PORT = process.env.PORT || 3000;`  . Если не задан порт, тогда будет `3000` по умолчанию.

gennady-bars commented on package.json:
```diff
@@ -22,9 +38,21 @@
+    "postcss-less": "^6.0.0",
+    "postcss-lit": "^0.5.0",
+    "stylelint": "^14.15.0",
+    "stylelint-config-standard": "^29.0.0",
```
Лучше задавать **фиксированные версии** зависимостям. Если нечаянно удалить `package-lock.json` и попытаться повторно установить зависимости с помощью команды `npm install`, то **npm** установит последние версии зависимостей, что может привести к ошибкам.

gennady-bars commented on package.json:
```diff
@@ -22,9 +38,21 @@
+    "eslint-config-airbnb-base": "15.0.0",
+    "eslint-plugin-import": "2.25.4",
+    "handlebars": "4.7.7",
+    "nanoid": "^4.0.0",
```
В `devDependencies` указывается то, что необходимо для сборки и разработки приложения, а в `dependencies` нужно указывать то, что необходимо непосредственно для работы приложения (**express, uuid**). Все линтеры, тесты и тд нужно только в `devDependencies` указывать, чтобы зря не раздувать кол-во пакетов в собранном для деплоя приложении.

gennady-bars commented on src/styles/variables.less:
```diff
@@ -0,0 +1,13 @@
+@box-shadow-color: #00000023;
+@background-color: #FBFBFB;
+@background-color-2: #EFEFEF;
+@background-color-3: #F8F8F8;
```
* Отлично, что цвета вынесены в `css`-константы

gennady-bars commented on src/pages/login/LoginPage.ts:
```diff
@@ -0,0 +1,89 @@
+      name: 'login',
+      events: {
+        focusin: (e) => {
+          inputValidation(e.target as HTMLInputElement);
```
```js

(e) => {
          inputValidation(e.target as HTMLInputElement);
        }
```

вот это повторяется в коде 34 раза. Значит, нужно вынести в отдельную функцию, типизировать 1 раз


```js

export const validationHandler = (e: Event) => {
  inputValidation(e.target as HTMLInputElement);
}


```


 и использовать в коде везде


```js

      events: {
        focusin: validationHandler,
        focusout: validationHandler,
      },


```


а теперь объект `events` повторяется 16 раз, значит, нужно вынести в константу и объект тоже


```js

export const validationHandler = (e: Event) => {
  inputValidation(e.target as HTMLInputElement);
};

export const eventsObject = {
  focusin: validationHandler,
  focusout: validationHandler,
};


```


и использовать в коде теперь только аккуратную строчку


```js

      events: eventsObject,


```


а еще лучше поместить вот этот `events: eventsObject,` прямо внутрь класса `InputFiled`, добавляя его внутрь пропсов, тогда не нужно будет дублировать 16 раз это в коде

Kosurij replied:
> 

Если написать так, то почему-то не навешиваются события. Т.е. я не жду объект событий, как пропсы,а сразу хардкожу его в инпут, но в таком случае события не добавляются. С любым другим полем - все ок.

![image](https://user-images.githubusercontent.com/62155518/204140636-896df940-2e9f-45fa-8efb-74145f503b6c.png)

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Работа проделана огромная


* `Readme` хорошо оформлен
* Проект задеплоен и отлично работает
* Отлично, что не забыли про `.gitignore`
* Отлично, что ограничиваете версию `node`
* Отлично, что фиксируете версии зависимостей (почти везде)
* Отлично, что нет `EOF` ошибок в гите
* Хорошая структура папок и файлов
* Добавлен `ESLint` и `Stylelint`
* Отлично, что цвета вынесены в `css`-константы
* Отлично, что не забываете про `alt` в тегах `img`
* Отлично, что отображаете текст ошибки под инпутом

но есть некоторые недочеты: 

* сейчас очищается валидация с инпута, если я начал что-то печатать в другом инпуте , а должна оставаться. Текс ошибки должен отображаться под инпутом, если уже прошла валидация, и инпут до сих пор невалидный. Скрин https://disk.yandex.ru/i/9Xzi3IgAsp3YOw 
* Страницы логина и регистрации тоже нужно обернуть в тег `main` по семантике. По семантике нужно помещать весь главный контент страницы в тег `main`.



**Можно лучше**


* При валидации лучше отображать **конкретный текст ошибки** под инпутом, чтобы было понятно, что пользователь делает не так и как это исправить. А то сейчас просто пишет, что `невалидный формат`, но что нужно конкретно, не пишет


Исправьте, пожалуйста, недочеты и работа будет принята. Пожалуйста, проверьте **работоспособность проекта** и наличие возможных ошибок в консоли браузера (кнопка `F12`) перед отправкой на ревью. 

Напоминаю, что работа может быть принята только после исправления всех критических замечаний `Нужно исправить`.

Удачного рефакторинга кода.

Commit:
Поле версия ноды перенесена в .nvmrc
Исправлена валидация.
События focusin/focusout сразу добавляются в классе InputField.
Убраны лишние логи и div.

Commit:
Скорректирована переменная PORT

Kosurij commented:
> **Можно лучше**
> 
> * При валидации лучше отображать **конкретный текст ошибки** под инпутом, чтобы было понятно, что пользователь делает не так и как это исправить. А то сейчас просто пишет, что `невалидный формат`, но что нужно конкретно, не пишет
>

Валидация планирую переделать, когда переедем на реакт: откажусь от варианта со стилизацией через псевдоэлемемнт, добавлю причину ошибки для каждого инпута.

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
